torch/_higher_order_ops/associative_scan.py

```diff
@@ -186,25 +191,37 @@ def associative_scan_functionalize(ctx, combine_fn, input, dim):
 
 
 @associative_scan_op.py_impl(torch._C._functorch.TransformType.Vmap)
-def associative_scan_batch_rule(interpreter, input, dim, combine_fn):
-    input_ = [get_unwrapped(x) for x in input]
-    input_bdims = [maybe_get_bdim(x) for x in input]
+def associative_scan_batch_rule(interpreter, combine_fn, input, dim):
+    input_bdims = [maybe_get_bdim(x) if is_batchedtensor(x) else None for x in input]
 
     batch_size = None
     for inp, bdim in zip(input, input_bdims):
         if bdim is not None:
-            batch_size = get_unwrapped(inp).shape[bdim]
+            batch_size = (
+                get_unwrapped(inp).shape[bdim]
+                if is_batchedtensor(inp)
+                else inp.shape[bdim]
+            )
 
     assert batch_size
+
+    if dim in input_bdims:
+        raise ValueError("Vmap in_dim may not conincide with dim of associative_scan")
```

Comment:
Is it ever possible for a user to hit this error?

Reply:
Yes, the user can hit this error if the call is invoked as
```
def associative_scan_fct(x):
    return associative_scan(add, x, 0, reverse=reverse)
 associative_scan1 = torch.vmap(associative_scan_fct, in_dims=0, out_dims=0)
```
The point here is that the user has full over what is the scan dimension and what is the vmap dimension. The two may coincide, which is what's prevented here.

Reply:
The dimension passed to scan inside the vmap is *different* from the in_dims in vmap. Inside the vmap, dimension 0 actually means "the 0th dimension  not including the vmapped dimension"

Reply:
Hmm, then I think something may be off here. This code snippet
```
x = torch.tile(
    torch.unsqueeze(
        torch.arange(
            0, 10, device=device, dtype=torch.float32, requires_grad=True
        ),
        0,
    ),
    (4, 1),
)
torch.compiler.reset()

def associative_scan_fct(x):
    return associative_scan(add, x, 0, reverse=reverse)

associative_scan1 = torch.compile(
    torch.vmap(associative_scan_fct, in_dims=1, out_dims=1), fullgraph=True
)

result1 = associative_scan1(x)
```
Calls the vmap with the input x of shape `4x10`, a `in_dims=1` and a scan dim of `0`. The result is that `input_bdims=[1]` in `associative_scan_batch_rule`. Now, when I change `in_dims=0`, then `input_bdims=[0]` and the scan dim is 0.